From 9a8defa99ed7aa4b0cbf9edbb9b10ab96dff08e6 Mon Sep 17 00:00:00 2001
From: Jose Dapena Paz <jdapena@igalia.com>
Date: Thu, 20 Aug 2020 17:02:27 +0000
Subject: [PATCH] GCC: SortByToken operators in PerFrameV8MemoryUsageData cannot be constexpr.

These operators fail to build in GCC as the mojo StructPtr operator-> is not constexpr either:

../../components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc:60:13: error: call to non-‘constexpr’ function ‘mojo::StructPtr<S>::Struct* mojo::StructPtr<S>::operator->() const [with S = blink::mojom::PerFrameV8MemoryUsageData; mojo::StructPtr<S>::Struct = blink::mojom::PerFrameV8MemoryUsageData]’
   60 |     return a->frame_token < b->frame_token;
      |             ^~
../../components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc: In member function ‘constexpr bool performance_manager::v8_memory::{anonymous}::SortByToken::operator()(const PerFrameUsagePtr&, const LocalFrameToken&)’:
../../components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc:65:13: error: call to non-‘constexpr’ function ‘mojo::StructPtr<S>::Struct* mojo::StructPtr<S>::operator->() const [with S = blink::mojom::PerFrameV8MemoryUsageData; mojo::StructPtr<S>::Struct = blink::mojom::PerFrameV8MemoryUsageData]’
   65 |     return a->frame_token < b.value();
      |             ^~

Bug: 819294
Change-Id: I4dbd8ee5f1f4b00b19a21777c4c8c93fdd7f6f16
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/2367038
Reviewed-by: François Doray <fdoray@chromium.org>
Commit-Queue: José Dapena Paz <jdapena@igalia.com>
Cr-Commit-Position: refs/heads/master@{#800190}
---

diff --git a/components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc b/components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc
index 6aa5995..a4279c8 100644
--- a/components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc
+++ b/components/performance_manager/v8_memory/v8_per_frame_memory_decorator.cc
@@ -55,13 +55,11 @@
 // Comparator that generates a strict total order of PerFrameUsagePtr's when
 // compared by their frame tokens.
 struct SortByToken {
-  constexpr bool operator()(const PerFrameUsagePtr& a,
-                            const PerFrameUsagePtr& b) {
+  bool operator()(const PerFrameUsagePtr& a, const PerFrameUsagePtr& b) {
     return a->frame_token < b->frame_token;
   }
 
-  constexpr bool operator()(const PerFrameUsagePtr& a,
-                            const blink::LocalFrameToken& b) {
+  bool operator()(const PerFrameUsagePtr& a, const blink::LocalFrameToken& b) {
     return a->frame_token < b.value();
   }
 };
